core/client/hdfs/src/main/java/alluxio/hadoop/LocalCacheFileSystem.java

```diff
@@ -148,9 +149,19 @@ public FSDataInputStream open(URIStatus status, int bufferSize) throws IOExcepti
       return mExternalFileSystem.open(HadoopUtils.toPath(new AlluxioURI(status.getPath())),
           bufferSize);
     }
-    return new FSDataInputStream(new HdfsFileInputStream(
-        new LocalCacheFileInStream(status, mAlluxioFileOpener, mCacheManager, mAlluxioConf),
-        statistics));
+    return new FSDataInputStream(new HdfsFileInputStream(new PositionReadFileInStream(
+        LocalCachePositionReader.create(mAlluxioConf, mCacheManager,
+            new CloseableSupplier<>(() -> {
+              try {
+                return new AlluxioHdfsPositionReader(mHadoopFileOpener.open(status),
+                    status.getLength());
+              } catch (IOException e) {
+                throw AlluxioRuntimeException.from(e);
+              }
+            }),
+            status, mAlluxioConf.getBytes(PropertyKey.USER_CLIENT_CACHE_PAGE_SIZE),
+            status.getCacheContext()),
+        status.getLength()), statistics));
```

Comment:
Please split the nested statements into multiple lines. This is not very easy to read.

Reply:
I am kind of debating between more readable vs more memory saving (less local variable)
This one is more complicated, will split to different lines